### ISSUE TITLE
Editor: IME input problem

### ISSUE THREAD

**Issue by minwe** [TryGhost/Ghost#9801]
### Issue Summary

![non-latin languages input problem in koenig editor](https://user-images.githubusercontent.com/4370770/44322775-99a0f880-a481-11e8-9b13-083bfed63679.gif)

As the screen recording above shown, I want to input `你好` using Chinese Pinyin Input method of Mac OS X on the start of line，I type `nihao`，then got `i好n`.

The first typed `n` is inserted after cursor.

### To Reproduce

1. Switch Input method to non-latin languages, e.g. Chinese Pinyin Input method, Japanese
2. Type some letter like `nihao`, expect to get `你好`, but get `i好n`

It is the bug of Mobiledoc-Kit, can reproduce on https://bustle.github.io/mobiledoc-kit/demo/, too.

### Technical details:

* OS: Mac OS X, v10.13.6
* Node Version: v8.11.1
* Ghost-CLI Version: 1.9.0
* Browser/OS:  Chrome/67.0.3396.99
* Database: sqlite3

**Comment by kevinansfield** [comment]
Hey @minwe 👋 Thanks for the report. Unfortunately this is a known issue, it's being tracked in the upstream mobiledoc-kit repo in https://github.com/bustle/mobiledoc-kit/issues/548 and I believe the underlying problem is similar to this issue https://github.com/bustle/mobiledoc-kit/issues/589.

**Comment by minwe** [comment]
Yes, this is same as https://github.com/bustle/mobiledoc-kit/issues/548.

It maybe caused by composition event (more about omposition event:  http://blog.evanyou.me/2014/01/03/composition-event/)

I'll check mobiledoc-kit source code later.

**Comment by kevinansfield** [comment]
It would be great if you were able to look. None of the core team use IME input so it's difficult to find and debug these issues.

https://github.com/bustle/mobiledoc-kit/issues/589 is a good read as I believe it's likely the root cause and has more info in it. Unless you are seeing the same buggy behaviour when not typing the first word in a blank paragraph?

**Comment by minwe** [comment]
@kevinansfield It always happens when typing the first word of paragraph, headline, list and other block element.

I'm looking at the issues you mentioned above, but have no idea so far.

**Comment by taking** [comment]
Korean has the same problem.

**Comment by S8Cloud** [comment]
It seems using IME in markdown section or title section works fine.  Hoping there is a option to disable Koeing Editor before it can work along with IME.

**Comment by buchijiang** [comment]
So when can it solved?

**Comment by vikaspotluri123** [comment]
@buchijiang whenever the upstream issue (https://github.com/bustle/mobiledoc-kit/issues/548) is solved - there's no explicit timeline publicly available

**Comment by buchijiang** [comment]
@vikaspotluri123 OK,i know and thanks for answer

**Comment by Windfarer** [comment]
There should be an alternative editor to this issue.

**Comment by soulteary** [comment]
> @buchijiang whenever the upstream issue ([bustle/mobiledoc-kit#548](https://github.com/bustle/mobiledoc-kit/issues/548)) is solved - there's no explicit timeline publicly available

Thanks.

TLDR;

But the mobiledoc-kit's **Maintainers** don't seem to merge it, so I created new repo for publish NPM repo: _@bugfix/mobiledoc-kit_ , source code here: https://github.com/soulteary/mobiledoc-kit

Update the **package.json** deps which in the ghost-admin project, and rebuild the assets with prod env, everything is okay.



![preview](https://user-images.githubusercontent.com/1500781/51799533-da851080-225c-11e9-87e5-4660c0354a0e.gif)

**Comment by miaowing** [comment]
When to fix it in master?

**Comment by taking** [comment]
Is it technology that can not be fixed?

**Comment by kevinansfield** [comment]
This is an issue in the upstream mobiledoc-kit library and the underlying problem is being tracked in multiple places on that repo, with the two most prominent issues being:

- https://github.com/bustle/mobiledoc-kit/issues/589
- https://github.com/bustle/mobiledoc-kit/issues/548

As such I'm going to close the issue here, any further discussion/PRs should be directed to the mobiledoc-kit project.

**Comment by xufangda** [comment]
> This is an issue in the upstream mobiledoc-kit library and the underlying problem is being tracked in multiple places on that repo, with the two most prominent issues being:
> 
> * [bustle/mobiledoc-kit#589](https://github.com/bustle/mobiledoc-kit/issues/589)
> * [bustle/mobiledoc-kit#548](https://github.com/bustle/mobiledoc-kit/issues/548)
> 
> As such I'm going to close the issue here, any further discussion/PRs should be directed to the mobiledoc-kit project.

As a blog platform, issues relating to editor is a core issue. This is not something like a development or dependency issue. This issues occurs at every paragraph when user write. Can you image how annoying it is everytime when you start a new paragraph, your first character jumps off.

It has been here for 2 years, and you don't resolve it, given the condition that one user has provided a solution after long time suffering. 

If you check out that user's blog, he has decided to leave ghost because of the attitude of the official team. 

This issue has caused so much trouble to all Asian language speakers. 

Very disappointed to the official team.

**Comment by kdyzm** [comment]
> > This is an issue in the upstream mobiledoc-kit library and the underlying problem is being tracked in multiple places on that repo, with the two most prominent issues being:
> > 
> > * [bustle/mobiledoc-kit#589](https://github.com/bustle/mobiledoc-kit/issues/589)
> > * [bustle/mobiledoc-kit#548](https://github.com/bustle/mobiledoc-kit/issues/548)
> > 
> > As such I'm going to close the issue here, any further discussion/PRs should be directed to the mobiledoc-kit project.
> 
> As a blog platform, issues relating to the editor is a core issue. This is not something like a development or dependency issue. This issue occurs in every paragraph when user writes.
> 
> Can you imagine how annoying every time when you start a new paragraph, your first character jumps off?
> 
> It has been here for 2 years, and you don't resolve it, given the condition that one user has provided a solution after a long time suffering.
> 
> If you check out that user's blog, he has decided to leave ghost because of the attitude of the official team.
> 
> This issue has caused so much trouble for all Asian language speakers.
> 
> Very disappointed with the official team.

In fact, what they want to say is that Asian users are not important, otherwise they would have solved the problem. After all, the users of this blogging platform are mainly Europeans and Americans. But I'm surprised that this bug has been resolved by some people, why don't they merge to master

**Comment by wuliaopage** [comment]
For anyone who wants to push this bug to be fixed, Please show your support underhttps://github.com/TryGhost/Ghost/issues/11460

**Comment by J-J-Chiarella** [comment]
I finally found this bug. I am not alone . . . and this is nearly a decade old.

The linked upstream issues are for a tool (bustle/mobiledoc-kit) that has not had a release in several years and no activity in two years.

It seems odd that Ghost Foundation is legally in Singapore (does someone have some connection with that place?) where many people and most elites use some form of Chinese, where this problem would come up. It strikes me as akin to chartering a company in Egypt with a product that cannot handle Arabic.

Is this a work in progress somewhere . . . or did people who rely on this just give up and move on? I am sensing the latter. Just seems odd is all.

**Comment by soulteary** [comment]
Reminded by this notification @J-J-Chiarella , seeing this information brings back quite a few thoughts.

I believe many people have fixed this issue before and managed to keep using it for a while. I did the same. To continue using it, I actually made some tweaks and put in extra effort. Back then, we didn't have the great AI tools we have now to quickly assist with debugging.

In 2020, I submitted a patched version to my personal repository and did not raise another pull request, because the one I had previously submitted was not merged by the official team, which was quite discouraging.

https://github.com/TryGhost/Admin/compare/main...soulteary:Ghost-Admin:master
https://github.com/TryGhost/mobiledoc-kit/compare/master...soulteary:mobiledoc-kit:master

**Comment by J-J-Chiarella** [comment]
Thank you for making the patch, @soulteary . I am sad to hear that it did not get accepted. Unfortunately, the internationalization of Ghost still has several tasks.

Yesterday, I stumbled upon some blog from years ago that compared WordPress with Ghost. The person was a Pole in Poland. He remarked on the longstanding and unresolved bugs that related to internationalization.

Maybe the situation with weaker internationalization in Ghost reflects the smallness and newness of Ghost. Or, maybe the team has different priorities. Maybe the core team wants to get ActivityPub working before it makes XYZ replaceable or before it engineers language switching on the interface for website visitors. These are just examples as part of my speculation.

I do not know if this information helps, but the problem manifests a bit differently for Korean hangul. In the example in the original post, I see *ni hao* become *i 好 n*. Korean hangul is itself an alphabet but this is not simple substitution as in Cyrillic. Korean hangul organizes letters into blocks (usually pre-composed in Unicode) for syllables. Imagine if *polo* were *po-lo* and *basket* were *bask-et* or whatever. 

The software should dynamically show a preview of the possible final output and then enter that code and move onto the next syllable whenever you move the cursor or type something impossible to keep in the same syllable. `안녕하세요` is `ㅇㅏㄴㄴㅕㅇㅎㅏㅅㅔㅇㅛ`. `ㅇ` is an empty place-holder or a final *ng* sound and `ㄴㄴ` can never appear consecutively in the same syllable. Therefore, the software sees input of `ㅇㅏㄴㄴ` and says, *Oh, the first syllable is `안` and then `ㄴ` is the beginning of a new syllable*.

With Ghost, it will never input a completed syllable and move on. The only workaround is to insert a plain character, maybe by typing ASCII punctuation such as `.`, and then writing, and then later deleting that initial `.`.
